Upgraded libmediascan

chincheta0815 opened the pull request:
Maybe it makes sense to implement the changes bit by bit via if-statements. If e.g. SunOS is tested it becomes default, same with FreeBSD. All others stay at the old-patched version until all are checked. This way we do not have to test on all platforms on ourselves as we do not have them all available...

fsbruva commented:
Once we get it working, we can ask Michael which way he'd prefer. He is basically the tester for the Linux and arm builds.

fsbruva commented:
@chincheta0815 
I am finally able to test locally. But - you changed the mode of EVERY SINGLE FILE TO BE EXECUTABLE (644 -> 755).  What is the reason for this change?

chincheta0815 commented:
@fsbruva 
Soryy. This seems to be an accident. I intentionally did that for my system but I did not want to upload that. I thought of git ignoring that change... I will have to check for the correct git option... In that case I feel like a noob. Sorry.

chincheta0815 commented:
@fsbruva 
I will reset that and push the update.

fsbruva commented:
Build of libmediascan-0.2 fails on FreeBSD 7.2, with system bundled gcc (4.2.1):
`libtool: compile:  cc -DHAVE_CONFIG_H -I. -I.. -Wall -I../include -I/usr/local/src/my_amd64_vendor/CPAN/build/include -I/usr/local/src/my_amd64_vendor/CPAN/build/include -fPIC -O3 -c mediascan_unix.c -o libmediascan_la-mediascan_unix.o
mediascan_unix.c: In function 'recurse_dir':
mediascan_unix.c:101: error: label 'out' used but not defined
mediascan_unix.c: At top level:
mediascan_unix.c:216: warning: data definition has no type or storage class
mediascan_unix.c:216: warning: type defaults to 'int' in declaration of 'closedir'
mediascan_unix.c:216: warning: parameter names (without types) in function declaration
mediascan_unix.c:219: error: expected identifier or '(' before 'if'
mediascan_unix.c:224: error: expected identifier or '(' before 'while'
mediascan_unix.c:232: warning: data definition has no type or storage class
mediascan_unix.c:232: warning: type defaults to 'int' in declaration of 'free'
mediascan_unix.c:232: warning: parameter names (without types) in function declaration
mediascan_unix.c:232: error: conflicting types for 'free'
/usr/include/stdlib.h:94: error: previous declaration of 'free' was here
mediascan_unix.c:234: error: expected '=', ',', ';', 'asm' or '__attribute__' before ':' token
mediascan_unix.c:236: error: expected identifier or '(' before '}' token
`

fsbruva commented:
Error occurs in both amd64 and i386. Problem is with this section of pre-processor directives, located at line 123 of mediascan_unix.c:
`#if defined(__APPLE__) || defined(__linux__)`
`      if (dp->d_type == DT_DIR) {`
`#elif defined(__sun__)`
`      if (PathIsDirectory(tmp_full_path)) {`
`#endif`

Since I don't have APPLE, linux or sun defined, there is a missing 'if' statement in my code by the time it is ready to be compiled. Shouldn't the code be:
`#if defined(__sun__) `
`      if (PathIsDirectory(tmp_full_path)) {`
`#else`
`      if (dp->d_type == DT_DIR) {`
`#endif` 

With the above change, the library is built. However, the Perl binding fails with:
`Running Mkbootstrap for Media::Scan ()
chmod 644 Scan.bs
rm -f blib/arch/auto/Media/Scan/Scan.so
LD_RUN_PATH="/usr/lib:/usr/local/lib" cc  -Wl,-shared -fPIC Scan.o  -o blib/arch/auto/Media/Scan/Scan.so /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libmediascan.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libavformat.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libavcodec.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libswscale.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libavutil.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libexif.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libjpeg.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libpng.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libgif.a /usr/local/src/my_amd64_vendor/CPAN/build/include/../lib/libdb.a      \
   -lm -lz -lintl       \

/usr/bin/ld: /usr/lib/crt1.o: relocation R_X86_64_32 can not be used when making a shared object; recompile with -fPIC
/usr/lib/crt1.o: could not read symbols: Bad value
gmake: *** [blib/arch/auto/Media/Scan/Scan.so] Error 1`

chincheta0815 commented:
Thanks for the info. 

Concerning the linux etc you are right. That seems reasonable and I will check and change that.

Concerning the second: Could you please look into the Makefile.PL of the perl binding and check whether there is a -shared lddflag at the end of the file? see:
https://github.com/chincheta0815/libmediascan/blob/6e1c502263fba27c5b6efc7742a73f2ac7a0e597/bindings/perl/Makefile.PL#L360

its just to check if its there. I never got that issue and just can suppose what it is. Seems to be frrebsd specfic up to now.

fsbruva commented:
> D_RUN_PATH="/usr/lib:/usr/local/lib" cc -Wl,-shared -fPIC Scan.o -o blib/arch/auto/Media/Scan/Scan.so 

Yes, the -shared code is being loaded. The error still occurs.

> Seems to be frrebsd specfic up to now.

Yes, but it's a build error that didn't exist with my branch before you made all your edits. libmediascan-0.1, as patched in my branch, builds without error. Therefore, I prefer we figure out where you introduced the error.

fsbruva commented:
Figured it out:
https://github.com/chincheta0815/libmediascan/blob/6e1c502263fba27c5b6efc7742a73f2ac7a0e597/bindings/perl/Makefile.PL#L361
Needs to be:
`$opts{LDDLFLAGS} = '-shared';`
Can you make that change and see if it still works for you?

fsbruva commented on CPAN/buildme.sh:
```diff
@@ -1396,11 +1396,15 @@ function build_ffmpeg {
-    
+  
+    # Disable h264 on SunOS as there are relocation error when linking to Media::Scan. 
+    if [ "$OS" = "SunOS" ] && [ "$FFMPEG_PREFIX" = "ffmpeg-3.4.1" ]; then
```
Once you make the requested change in the bindings/perl/Makefile.PL ('-W1,-shared' --> '-shared'), can you revert this to see if it links properly?

chincheta0815 commented:
Okay, thanks for the info. 

The build issue that made me introduce the LDflag was related to h264. When I compiled without -shared I could not use ffmpeg with h264, with h264 I had to use -shared. ffmpegs faqs therefore say that one should use -Bsymbolic when linking static libs into a shared lib which did not work (https://ffmpeg.org/platform.html#toc-Advanced-linking-configuration). I searched a little bit and found shared to be working.

As it seems to work without -Wl it seems to be compilerversion or platform specific. I will check if it works for me. One all is done I will remove that h264 comment you mentioned earlier.

Sorry about the mees concerning the file permissions...

chincheta0815 commented on CPAN/buildme.sh:
```diff
@@ -463,7 +463,7 @@ elif [ "$OS" = "SunOS" ]; then
         exit
     fi 
-    export MAKE=/usr/bin/gmake
+    export MAKE="/usr/bin/gmake -j4"
```
Does j4 make sense here. When one simple wants to build on 1 core or a different number...

fsbruva replied:
It only makes sense if the cpu is quad core (Or dual w/threads). Do you have a build problem if you omit the -j?

fsbruva commented:
No problem. I still have to add the db version detection code before I bother Michael with this.

Here is the site that pointed me to the removal of -W1... https://stackoverflow.com/questions/36070158/linking-errors-regarding-fpic-even-though-all-sources-are-compiled-with-fpic 

I am using gcc 4.2.1 and 4.8.5.

chincheta0815 commented:
I changed the build script to not compile ffmpeg with asm on solaris as there seems to be an issues there. Everything works well except including h264 with asm. I suppose there might be a problem with gcc 5 which is standard on illumos.

I also tried if there are problems with deprecated stuff in ffmpeg. Therefore I cleaned up libmediascan a little bit to get rid of warnings etc. I am not yet done with that, but i compiles well now.

fsbruva commented:
Excellent! Thank you for the help with this. I will not be able to do the merge for a few days. Let me know when it's satisfactory, and I can squash/merge.

Regarding asm - I had the same problems with FreeBSD, but only when building in a jail, for some reason. Same compiler, same build chain, same code from git. And the h264_cabac would fail every time. Part of the issue might be the archaic gcc that I am using. FFmpeg has documented/known issues building on gcc 4.2.1, [FFmpeg ticket ](https://trac.ffmpeg.org/ticket/3970). Should therefore set the minimum compiler version (or at least warn) in line 174?

fsbruva commented:
Did you fork libmediascan to make all those changes you did? That might be a good idea, so if Michael or Andy want to see the changes with markup. Edit: Nevermind. I see you did.

chincheta0815 commented:
I do not know if a warning makes sense as I was using gcc 5.2 which is newer than the one you are using. It seems to be an annoying overall bug, but I am not able to solve that due to lacking knowledge...

Yep, there is a fork.

chincheta0815 commented:
@fsbruva after a long time not working in this, I have found the clue how to compile with native perl on solaris. I had to add "LD=$GCC" for Media::Scan and for XML::Parser as these have problems with the sun linker in /usr/css/bin...
The LD=$GCC might be a good idea for all makes, but I could not test that due to only having OmniOS...
I also upgraded ffmpeg to version 4.0.

Could you check I that works once in a while?
Then we might push to Michael...

fsbruva commented:
I will give this a try on my Ubuntu and FreeBSD boxes soon. I made some comments in your libmediascan repository - your recent commit added a bunch of trailing whitespace back in (result.c).

chincheta0815 commented:
Okay, thanks for the comments. I tried to find them. Did you already push them? Or where are they (early in the morning here ;o) )
Once everything works I will also (again) work on my whitespaces...

chincheta0815 commented:
Sorry. Really early.... I found the comments.
Well, you are so right. Whitespaces will be killed this afternoon. And the other changes concerning your comments will be undone.... I do no know what happened there.

fsbruva commented:
A few issues: 
1. The libmediascan-0.2.tar.bz doesn't match your repository, so the build fails for me. 
2. One of your changes (re) breaks the build in FreeBSD. It's the -Bsymbolic flag in the libmediascan Makefile.pl (required by FFmpeg in FreeBSD and Linux). I will re-create a OmniOS virtual machine, and replicate the success I had in February with getting it to build with the -Bsymbolic flag set.
3. Something you did in the configure.ac stuff REALLY breaks the final linker: /usr/bin/ld can't find the -lintl library, even though it's installed. I will have to investigate and track this down.

fsbruva commented:
Oh good - you had the same thought as me... to make the LDDFlags conditional on OS. Excellent! :-)

Did you revert any changes in configure.ac? I haven't had a chance to track them down. If you found a way to get asm to work - then I don't want to take that capability away. I'd rather find the "right" solution, rather than the easy one.

I have already been meaning to test if all the explicit CC=$GCC calls I put in could be replaced with a single `export CC=$GCC`. If that works, then we could setup a conditional `export LD=$GCC` for Sun/Solaris. Thoughts?

chincheta0815 commented:
Concerning the changes I did not revert anything. all I did was adapting some compatibility issues due to a higher version in the whole automake stuff. I then changed the routine and added the conditionals for sun etc. I did not (at least on purpose) touch -lintl. On my system it works well. Maybe you ran into the thing, that you have to run autogen.sh before you are able to compile anything...

Concerning the CC=$GCC it sounds reasonable to me - especially on SunOS/IllumOS as there usually the sun compilers might be used. So CC=$GCC (and $CXX, etc.) is a good thing there. Concerning LD=$GCC it is a little bit more tricky: SunOS/Illumos reacts quite picky there. There seems to happen always a call to /usr/bin/ld which then changes behavior respecting LD_ALTEXEC=, ignoring LD=, although the meaning is the same... Weird. I have not quite understood that.

Maybe another idea would be to set the PATH for SunOS/Illumos prepending /usr/gnu/bin as there all gnu-tools reside that are needed by the buildme for a correct build...

chincheta0815 commented:
I just checked your idea of having global $GCC, etc.
Well, they are already in there. Or did I not get your idea correctly?

Settings those solaris specific stuff globally breaks others things, no idea why. So leaving it as is seems okay - at least for solaris.

By the way: I got the issue in the Makefile.PL for libmediascan. I just had to use "gld" directly with different flags (no -Wl). Then it compiles. I added that to the else-statement. Now ffmpeg can be used with ASM support.

The thing here is that SunOS uses suns ld, ALWAYS. So it ignores $LD, but reacts on $LD_ALTEXEC. Sun's LD reads $LD_ALTEXEC and then invokes it. Strange but seems to work...

Settings that globally breaks e.g. ICU...

fsbruva commented:
Nevermind. I got it figured it out.

Can you please accept https://github.com/chincheta0815/libmediascan/pull/4 ?

I get the following error on FreeBSD:
`/usr/bin/ld: cannot find -lintl
cc: error: linker command failed with exit code 1 (use -v to see invocation)
gmake: *** [Makefile:508: blib/arch/auto/Media/Scan/Scan.so] Error 1
make failed, aborting`

Even though ldconfig -r returns: 
`        188:-lintl.8 => /usr/local/lib/libintl.so.8`

The issue is the includes directories are a pain to set when libintl is not in the base. I need the knob enabled.

fsbruva commented:
I merged this PR into https://github.com/fsbruva/slimserver-vendor/tree/chincheta0815-upgraded_libmediascan

Please submit a new PR for the next updates. I had to manually merge the changes from mherger's public/7.9.
